heft-plugins/heft-lint-plugin/src/SarifFormatter.ts

```diff
@@ -0,0 +1,325 @@
+      }
+    }
+  };
+
+  const sarifLog: ISarifLog = {
+    version: SARIF_VERSION,
+    $schema: SARIF_INFORMATION_URI,
+    runs: [sarifRun]
+  };
+
+  let executionSuccessful: boolean = true;
+  let currentArtifactIndex: number = 0;
+  let currentRuleIndex: number = 0;
+
+  for (const result of results) {
+    const { filePath } = result;
+    const fileUrl: string = path.relative(buildFolderPath, filePath).toString();
+    let sarifFile: ISarifFile | undefined = sarifFiles.get(filePath);
+
+    if (sarifFile === undefined) {
+      sarifFile = {
+        location: {
+          uri: fileUrl
+        }
+      };
+      sarifFiles.set(filePath, sarifFile);
+    }
+
+    sarifArtifactIndices.set(sarifFile, currentArtifactIndex++);
+    const containsSuppressedMessages: boolean =
+      result.suppressedMessages && result.suppressedMessages.length > 0;
+    const messages: IMessage[] =
+      containsSuppressedMessages && !ignoreSuppressed
+        ? [...result.messages, ...result.suppressedMessages]
+        : result.messages;
+
+    for (const message of messages) {
+      const level: string = message.fatal || message.severity === 2 ? 'error' : 'warning';
+      const physicalLocation: ISarifPhysicalLocation = {
+        artifactLocation: {
+          uri: fileUrl,
+          index: sarifArtifactIndices.get(sarifFile)
+        }
+      };
+
+      const sarifRepresentation: ISarifRepresentation = {
+        level,
+        message: {
+          text: message.message
+        },
+        locations: [
+          {
+            physicalLocation
+          }
+        ]
+      };
+
+      if (message.ruleId) {
+        sarifRepresentation.ruleId = message.ruleId;
+
+        if (rulesMeta && sarifRules.get(message.ruleId) === undefined) {
+          const meta: TEslint.Rule.RuleMetaData = rulesMeta[message.ruleId];
+
+          // An unknown ruleId will return null. This check prevents unit test failure.
+          if (meta) {
+            sarifRuleIndices.set(message.ruleId, currentRuleIndex++);
+
+            if (meta.docs) {
+              // Create a new entry in the rules dictionary.
+              const shortDescription: string = meta.docs.description ?? '';
+
+              const sarifRule: ISarifRule = {
+                id: message.ruleId,
+                helpUri: meta.docs.url,
+                properties: {
+                  category: meta.docs.category
+                },
+                shortDescription: {
+                  text: shortDescription
+                }
+              };
+              sarifRules.set(message.ruleId, sarifRule);
+              // Some rulesMetas do not have docs property
+            } else {
+              sarifRules.set(message.ruleId, {
+                id: message.ruleId,
+                helpUri: 'Please see details in message',
+                properties: {
+                  category: 'No category provided'
+                }
+              });
+            }
+          }
+        }
```

Comment:
Doing this for every instance of every violated rule? You should probably create the list of rules once, and then go through the files and messages, looking up the index of each rule.

See my suggestion about artifact indices, you can do something similar with tracking rule indices.